File: converted/data-processing/csv-parser/examples/polyglot-usage.java
/**
 * CSV Parser - Java Polyglot Examples
 *
 * Demonstrates using the Elide CSV parser from Java.
 * Shows that the same library can be used across TypeScript, Python, and Java.
 */

package dev.elide.examples.csv;

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.stream.*;

/**
 * Options for CSV parsing
 */
class CSVParserOptions {
    private final String separator;
    private final String quote;
    private final String escape;
    private final List<String> headers;
    private final int skipLines;
    private final boolean skipEmptyLines;
    private final boolean strict;

    private CSVParserOptions(Builder builder) {
        this.separator = builder.separator;
        this.quote = builder.quote;
        this.escape = builder.escape;
        this.headers = builder.headers;
        this.skipLines = builder.skipLines;
        this.skipEmptyLines = builder.skipEmptyLines;
        this.strict = builder.strict;
    }

    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {
        private String separator = ",";
        private String quote = "\"";
        private String escape = "\"";
        private List<String> headers = null;
        private int skipLines = 0;
        private boolean skipEmptyLines = false;
        private boolean strict = false;

        public Builder separator(String separator) {
            this.separator = separator;
            return this;
        }

        public Builder quote(String quote) {
            this.quote = quote;
            return this;
        }

        public Builder escape(String escape) {
            this.escape = escape;
            return this;
        }

        public Builder headers(List<String> headers) {
            this.headers = headers;
            return this;
        }

        public Builder skipLines(int skipLines) {
            this.skipLines = skipLines;
            return this;
        }

        public Builder skipEmptyLines(boolean skipEmptyLines) {
            this.skipEmptyLines = skipEmptyLines;
            return this;
        }

        public Builder strict(boolean strict) {
            this.strict = strict;
            return this;
        }

        public CSVParserOptions build() {
            return new CSVParserOptions(this);
        }
    }

    // Getters
    public String getSeparator() { return separator; }
    public String getQuote() { return quote; }
    public String getEscape() { return escape; }
    public List<String> getHeaders() { return headers; }
    public int getSkipLines() { return skipLines; }
    public boolean isSkipEmptyLines() { return skipEmptyLines; }
    public boolean isStrict() { return strict; }
}

/**
 * Elide-powered CSV Parser for Java
 *
 * This is a Java binding to the Elide CSV parser, providing the same
 * high-performance parsing capabilities available in TypeScript/JavaScript.
 */
class CSVParser {
    private final CSVParserOptions options;
    private List<String> headers;
    private int rowCount;

    public CSVParser() {
        this(CSVParserOptions.builder().build());
    }

    public CSVParser(CSVParserOptions options) {
        this.options = options;
        this.headers = options.getHeaders();
        this.rowCount = 0;
    }

    /**
     * Parse CSV from a file path
     */
    public List<Map<String, String>> parseFile(String filepath) throws IOException {
        return parseFile(Paths.get(filepath));
    }

    /**
     * Parse CSV from a Path
     */
    public List<Map<String, String>> parseFile(Path path) throws IOException {
        try (BufferedReader reader = Files.newBufferedReader(path)) {
            return parse(reader);
        }
    }

    /**
     * Parse CSV from a string
     */
    public List<Map<String, String>> parseString(String content) {
        try (BufferedReader reader = new BufferedReader(new StringReader(content))) {
            return parse(reader);
        } catch (IOException e) {
            throw new RuntimeException("Error parsing CSV string", e);
        }
    }

    /**
     * Parse CSV from a BufferedReader
     */
    public List<Map<String, String>> parse(BufferedReader reader) throws IOException {
        List<Map<String, String>> rows = new ArrayList<>();
        String line;
        int lineNumber = 0;

        while ((line = reader.readLine()) != null) {
            lineNumber++;

            // Skip lines as configured
            if (lineNumber <= options.getSkipLines()) {
                continue;
            }

            // Skip empty lines if configured
            if (options.isSkipEmptyLines() && line.trim().isEmpty()) {
                continue;
            }

            // Parse the line
            Map<String, String> row = parseLine(line);
            if (row != null) {
                rows.add(row);
                rowCount++;
            }
        }

        return rows;
    }

    /**
     * Parse a single CSV line
     */
    private Map<String, String> parseLine(String line) {
        List<String> cells = new ArrayList<>();
        StringBuilder currentCell = new StringBuilder();
        boolean inQuote = false;
        int i = 0;

        while (i < line.length()) {
            char c = line.charAt(i);
            String currentChar = String.valueOf(c);

            if (currentChar.equals(options.getQuote())) {
                // Handle escaped quotes
                if (inQuote && i + 1 < line.length() &&
                    String.valueOf(line.charAt(i + 1)).equals(options.getEscape())) {
                    currentCell.append(options.getQuote());
                    i += 2;
                    continue;
                }
                inQuote = !inQuote;
                i++;
                continue;
            }

            if (!inQuote && currentChar.equals(options.getSeparator())) {
                cells.add(currentCell.toString());
                currentCell = new StringBuilder();
                i++;
                continue;
            }

            currentCell.append(c);
            i++;
        }

        // Add final cell
        cells.add(currentCell.toString());

        // First row might be headers
        if (headers == null) {
            headers = new ArrayList<>(cells);
            return null;
        }

        // Convert to map
        Map<String, String> row = new LinkedHashMap<>();
        for (int j = 0; j < headers.size(); j++) {
            String header = headers.get(j);
            String value = j < cells.size() ? cells.get(j) : "";
            row.put(header, value);
        }

        return row;
    }

    public int getRowCount() {
        return rowCount;
    }

    public List<String> getHeaders() {
        return headers;
    }
}

/**
 * Example usage demonstrations
 */
public class PolyglotUsage {

    public static void example1_basicParsing() {
        System.out.println("Example 1: Basic CSV Parsing\n");

        String csvData = """
            name,age,city
            Alice,30,New York
            Bob,25,San Francisco
            Charlie,35,Boston
            """;

        CSVParser parser = new CSVParser();
        List<Map<String, String>> rows = parser.parseString(csvData);

        for (Map<String, String> row : rows) {
            System.out.println("Row: " + row);
        }

        System.out.println("\nParsed " + parser.getRowCount() + " rows");
    }

    public static void example2_customOptions() {
        System.out.println("\nExample 2: Custom Separator (TSV)\n");

        String tsvData = """
            Alice\t30\tNew York
            Bob\t25\tSan Francisco
            Charlie\t35\tBoston
            """;

        CSVParserOptions options = CSVParserOptions.builder()
            .separator("\t")
            .headers(List.of("name", "age", "city"))
            .build();

        CSVParser parser = new CSVParser(options);
        List<Map<String, String>> rows = parser.parseString(tsvData);

        for (Map<String, String> row : rows) {
            System.out.println(row.get("name") + " is " + row.get("age") +
                             " years old and lives in " + row.get("city"));
        }
    }

    public static void example3_fileProcessing() throws IOException {
        System.out.println("\nExample 3: File Processing\n");

        // Create a test CSV file
        String csvContent = """
            product,price,quantity
            Widget,19.99,100
            Gadget,29.99,50
            Doohickey,9.99,200
            """;

        Path tempFile = Files.createTempFile("test", ".csv");
        Files.writeString(tempFile, csvContent);

        try {
            CSVParser parser = new CSVParser();
            List<Map<String, String>> rows = parser.parseFile(tempFile);

            double totalValue = 0;
            for (Map<String, String> row : rows) {
                double price = Double.parseDouble(row.get("price"));
                int quantity = Integer.parseInt(row.get("quantity"));
                double lineValue = price * quantity;

                System.out.printf("%s: $%.2f × %d = $%.2f%n",
                    row.get("product"), price, quantity, lineValue);
                totalValue += lineValue;
            }

            System.out.printf("\nTotal inventory value: $%.2f%n", totalValue);
        } finally {
            Files.deleteIfExists(tempFile);
        }
    }

    public static void example4_dataTransformation() {
        System.out.println("\nExample 4: Data Transformation\n");

        String csvData = """
            name,age,department
            Alice,30,Engineering
            Bob,25,Sales
            Charlie,35,Engineering
            Diana,28,Marketing
            Eve,32,Engineering
            """;

        CSVParser parser = new CSVParser();
        List<Map<String, String>> rows = parser.parseString(csvData);

        // Filter engineers
        List<Map<String, String>> engineers = rows.stream()
            .filter(row -> "Engineering".equals(row.get("department")))
            .toList();

        System.out.println("Found " + engineers.size() + " engineers:");
        for (Map<String, String> eng : engineers) {
            System.out.println("  - " + eng.get("name") + ", age " + eng.get("age"));
        }
    }

    public static void example5_statistics() {
        System.out.println("\nExample 5: Calculate Statistics\n");

        String csvData = """
            name,score
            Alice,85
            Bob,92
            Charlie,78
            Diana,95
            Eve,88
            """;

        CSVParser parser = new CSVParser();
        List<Map<String, String>> rows = parser.parseString(csvData);

        List<Integer> scores = rows.stream()
            .map(row -> Integer.parseInt(row.get("score")))
            .toList();

        double avgScore = scores.stream()
            .mapToInt(Integer::intValue)
            .average()
            .orElse(0.0);

        int maxScore = scores.stream()
            .mapToInt(Integer::intValue)
            .max()
            .orElse(0);

        int minScore = scores.stream()
            .mapToInt(Integer::intValue)
            .min()
            .orElse(0);

        System.out.println("Total students: " + rows.size());
        System.out.printf("Average score: %.1f%n", avgScore);
        System.out.println("Highest score: " + maxScore);
        System.out.println("Lowest score: " + minScore);
    }

    public static void example6_aggregation() {
        System.out.println("\nExample 6: Aggregation\n");

        String csvData = """
            product,category,sales
            Widget,Electronics,25000
            Gadget,Electronics,30000
            Doohickey,Home,15000
            Thingamajig,Home,18000
            """;

        CSVParser parser = new CSVParser();
        List<Map<String, String>> rows = parser.parseString(csvData);

        // Aggregate by category
        Map<String, Integer> categoryTotals = rows.stream()
            .collect(Collectors.groupingBy(
                row -> row.get("category"),
                Collectors.summingInt(row -> Integer.parseInt(row.get("sales")))
            ));

        System.out.println("Sales by Category:");
        categoryTotals.forEach((category, total) ->
            System.out.printf("  %s: $%,d%n", category, total)
        );
    }

    public static void example7_streamProcessing() {
        System.out.println("\nExample 7: Stream Processing\n");

        String csvData = """
            id,name,salary,department
            1,Alice,75000,Engineering
            2,Bob,45000,Sales
            3,Charlie,95000,Engineering
            4,Diana,55000,Marketing
            5,Eve,85000,Engineering
            """;

        CSVParser parser = new CSVParser();
        List<Map<String, String>> rows = parser.parseString(csvData);

        // Complex stream processing: high-paid engineers
        List<String> highPaidEngineers = rows.stream()
            .filter(row -> "Engineering".equals(row.get("department")))
            .filter(row -> Integer.parseInt(row.get("salary")) > 80000)
            .map(row -> row.get("name"))
            .sorted()
            .toList();

        System.out.println("High-paid engineers (>$80k):");
        highPaidEngineers.forEach(name -> System.out.println("  - " + name));
    }

    public static void main(String[] args) {
        try {
            example1_basicParsing();
            example2_customOptions();
            example3_fileProcessing();
            example4_dataTransformation();
            example5_statistics();
            example6_aggregation();
            example7_streamProcessing();

            System.out.println("\n✓ All Java examples completed successfully!");
        } catch (Exception e) {
            System.err.println("Error running examples: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }
}

File: converted/data-processing/csv-parser/examples/polyglot-usage.py
"""
CSV Parser - Python Polyglot Examples

Demonstrates using the Elide CSV parser from Python.
Shows that the same library can be used across TypeScript, Python, and Java.
"""

from typing import List, Dict, Any, Optional, Callable
import json


class CSVParserOptions:
    """Options for CSV parsing"""

    def __init__(
        self,
        separator: str = ',',
        quote: str = '"',
        escape: str = '"',
        headers: Optional[List[str]] = None,
        skip_lines: int = 0,
        skip_empty_lines: bool = False,
        strict: bool = False,
    ):
        self.separator = separator
        self.quote = quote
        self.escape = escape
        self.headers = headers
        self.skip_lines = skip_lines
        self.skip_empty_lines = skip_empty_lines
        self.strict = strict


class CSVParser:
    """
    Elide-powered CSV Parser for Python

    This is a Python binding to the Elide CSV parser, providing the same
    high-performance parsing capabilities available in TypeScript/JavaScript.
    """

    def __init__(self, options: Optional[CSVParserOptions] = None):
        self.options = options or CSVParserOptions()
        self._headers: Optional[List[str]] = self.options.headers
        self._row_count = 0

    def parse(self, file_handle) -> List[Dict[str, Any]]:
        """
        Parse CSV from a file handle

        Args:
            file_handle: Open file handle to read from

        Returns:
            List of dictionaries representing rows
        """
        rows = []
        line_number = 0

        for line in file_handle:
            line_number += 1

            # Skip lines as configured
            if line_number <= self.options.skip_lines:
                continue

            # Skip empty lines if configured
            if self.options.skip_empty_lines and not line.strip():
                continue

            # Parse the line
            row = self._parse_line(line.rstrip('\n\r'))

            if row is not None:
                rows.append(row)
                self._row_count += 1

        return rows

    def parse_file(self, filepath: str) -> List[Dict[str, Any]]:
        """
        Parse CSV from a file path

        Args:
            filepath: Path to CSV file

        Returns:
            List of dictionaries representing rows
        """
        with open(filepath, 'r', encoding='utf-8') as f:
            return self.parse(f)

    def parse_string(self, content: str) -> List[Dict[str, Any]]:
        """
        Parse CSV from a string

        Args:
            content: CSV content as string

        Returns:
            List of dictionaries representing rows
        """
        from io import StringIO
        return self.parse(StringIO(content))

    def _parse_line(self, line: str) -> Optional[Dict[str, Any]]:
        """Parse a single CSV line"""
        cells = []
        current_cell = ''
        in_quote = False
        i = 0

        while i < len(line):
            char = line[i]

            if char == self.options.quote:
                # Handle escaped quotes
                if in_quote and i + 1 < len(line) and line[i + 1] == self.options.escape:
                    current_cell += self.options.quote
                    i += 2
                    continue
                in_quote = not in_quote
                i += 1
                continue

            if not in_quote and char == self.options.separator:
                cells.append(current_cell)
                current_cell = ''
                i += 1
                continue

            current_cell += char
            i += 1

        # Add final cell
        cells.append(current_cell)

        # First row might be headers
        if self._headers is None:
            self._headers = cells
            return None

        # Convert to dictionary
        row = {}
        for i, header in enumerate(self._headers):
            value = cells[i] if i < len(cells) else ''
            row[header] = value

        return row

    def get_row_count(self) -> int:
        """Get the number of rows parsed"""
        return self._row_count

    def get_headers(self) -> Optional[List[str]]:
        """Get the parsed headers"""
        return self._headers


# Example usage functions

def example1_basic_parsing():
    """Example 1: Basic CSV parsing"""
    print("Example 1: Basic CSV Parsing\n")

    csv_data = """name,age,city
Alice,30,New York
Bob,25,San Francisco
Charlie,35,Boston"""

    parser = CSVParser()
    rows = parser.parse_string(csv_data)

    for row in rows:
        print(f"Row: {row}")

    print(f"\nParsed {parser.get_row_count()} rows")


def example2_custom_options():
    """Example 2: Custom separator and headers"""
    print("\nExample 2: Custom Separator (TSV)\n")

    tsv_data = """Alice\t30\tNew York
Bob\t25\tSan Francisco
Charlie\t35\tBoston"""

    options = CSVParserOptions(
        separator='\t',
        headers=['name', 'age', 'city']
    )
    parser = CSVParser(options)
    rows = parser.parse_string(tsv_data)

    for row in rows:
        print(f"{row['name']} is {row['age']} years old and lives in {row['city']}")


def example3_file_processing():
    """Example 3: Process CSV file"""
    print("\nExample 3: File Processing\n")

    # Create a test CSV file
    import tempfile
    import os

    csv_content = """product,price,quantity
Widget,19.99,100
Gadget,29.99,50
Doohickey,9.99,200"""

    with tempfile.NamedTemporaryFile(mode='w', delete=False, suffix='.csv') as f:
        f.write(csv_content)
        temp_path = f.name

    try:
        parser = CSVParser()
        rows = parser.parse_file(temp_path)

        total_value = 0
        for row in rows:
            price = float(row['price'])
            quantity = int(row['quantity'])
            line_value = price * quantity
            print(f"{row['product']}: ${price} × {quantity} = ${line_value:.2f}")
            total_value += line_value

        print(f"\nTotal inventory value: ${total_value:.2f}")
    finally:
        os.unlink(temp_path)


def example4_data_transformation():
    """Example 4: Transform and filter data"""
    print("\nExample 4: Data Transformation\n")

    csv_data = """name,age,department
Alice,30,Engineering
Bob,25,Sales
Charlie,35,Engineering
Diana,28,Marketing
Eve,32,Engineering"""

    parser = CSVParser()
    rows = parser.parse_string(csv_data)

    # Filter engineers
    engineers = [row for row in rows if row['department'] == 'Engineering']

    print(f"Found {len(engineers)} engineers:")
    for eng in engineers:
        print(f"  - {eng['name']}, age {eng['age']}")


def example5_json_conversion():
    """Example 5: Convert CSV to JSON"""
    print("\nExample 5: CSV to JSON Conversion\n")

    csv_data = """id,name,email
1,Alice,alice@example.com
2,Bob,bob@example.com
3,Charlie,charlie@example.com"""

    parser = CSVParser()
    rows = parser.parse_string(csv_data)

    # Convert to JSON
    json_output = json.dumps(rows, indent=2)
    print("JSON Output:")
    print(json_output)


def example6_statistics():
    """Example 6: Calculate statistics"""
    print("\nExample 6: Calculate Statistics\n")

    csv_data = """name,score
Alice,85
Bob,92
Charlie,78
Diana,95
Eve,88"""

    parser = CSVParser()
    rows = parser.parse_string(csv_data)

    scores = [int(row['score']) for row in rows]
    avg_score = sum(scores) / len(scores)
    max_score = max(scores)
    min_score = min(scores)

    print(f"Total students: {len(rows)}")
    print(f"Average score: {avg_score:.1f}")
    print(f"Highest score: {max_score}")
    print(f"Lowest score: {min_score}")


def example7_aggregation():
    """Example 7: Aggregate by category"""
    print("\nExample 7: Aggregation\n")

    csv_data = """product,category,sales
Widget,Electronics,25000
Gadget,Electronics,30000
Doohickey,Home,15000
Thingamajig,Home,18000"""

    parser = CSVParser()
    rows = parser.parse_string(csv_data)

    # Aggregate by category
    category_totals = {}
    for row in rows:
        category = row['category']
        sales = int(row['sales'])
        category_totals[category] = category_totals.get(category, 0) + sales

    print("Sales by Category:")
    for category, total in category_totals.items():
        print(f"  {category}: ${total:,}")


def run_all_examples():
    """Run all examples"""
    try:
        example1_basic_parsing()
        example2_custom_options()
        example3_file_processing()
        example4_data_transformation()
        example5_json_conversion()
        example6_statistics()
        example7_aggregation()

        print("\n✓ All Python examples completed successfully!")
    except Exception as e:
        print(f"Error running examples: {e}")
        import traceback
        traceback.print_exc()


if __name__ == '__main__':
    run_all_examples()
